WIP: SOLR-13150

moshebla opened the pull request.

gus-asf commented on solr/core/src/java/org/apache/solr/cloud/api/collections/CategoryRoutedAlias.java:
```diff
@@ -125,13 +146,13 @@ String buildCollectionNameFromValue(String value) {
 
   @Override
   public String createCollectionsIfRequired(AddUpdateCommand cmd) {
+    assert !getCollectionList(this.parsedAliases).contains(buildCollectionNameFromValue(String.valueOf(cmd.getSolrInputDocument().getFieldValue(getRouteField()))));
```
This assert is not correct... the method is createCollections--if--required. meaning it might not be required (because it might already exist) I'll remove this.

moshebla replied:
Oh sorry I missed that subtle nuance.

gus-asf commented on solr/core/src/java/org/apache/solr/cloud/api/collections/CategoryRoutedAlias.java:
```diff
@@ -117,15 +117,15 @@ public void validateRouteValue(AddUpdateCommand cmd) throws SolrException {
-    if (colls.stream()
+    if (cols.stream()
         .filter(x -> !x.contains(UNINITIALIZED)).count() >= Integer.valueOf(maxCardinality)) {
-      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "max cardinality can not be exceeded for a Category Routed Alias: " + maxCardinality);
```
Actually, Bad request is better. The server is not broken, it is correctly refusing data that it is configured to refuse. I'll tweak

gus-asf commented on solr/core/src/java/org/apache/solr/cloud/api/collections/CategoryRoutedAlias.java:
```diff
@@ -128,7 +128,8 @@ public void validateRouteValue(AddUpdateCommand cmd) throws SolrException {
     if (cols.stream()
         .filter(x -> !x.contains(UNINITIALIZED)).count() >= Integer.valueOf(maxCardinality)) {
-      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "max cardinality can not be exceeded for a Category Routed Alias: " + maxCardinality);
+      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Max cardinality " + maxCardinality
```
Yeah this message is clearer and friendlier :).

moshebla commented:
Rebased on SOLR-13131.
Updated the added test to include maxCardinality in the createAlias cmd, since it is now a required param and caused the newly added tests to fail.

gus-asf commented:
Thanks, merged via pull locally with a few fixups. Now available on branch solr-13131
